pkg/controllermanager/controller/cloudprofile/kubernetesversion/kubernetes_version.go

```diff
@@ -0,0 +1,164 @@
+			version = helper.SetSupportedClassification(version)
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = version
+		} else if version.Classification == nil {
+			// deprecate
+			deprecatedVersion, err := deprecateKubernetesVersion(logger, config, profile, version, *currentSemverVersion, hasSupportedMinorVersion, isLatestPatchVersion)
+			if err != nil {
+				return false, nil, err
+			}
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = *deprecatedVersion
+			continue
+		} else if !hasSupportedMinorVersion && (*version.Classification == gardencorev1beta1.ClassificationPreview || *version.Classification == gardencorev1beta1.ClassificationSupported) {
+			deprecatedVersion, err := deprecateKubernetesVersion(logger, config, profile, version, *currentSemverVersion, false, isLatestPatchVersion)
+			if err != nil {
+				return false, nil, err
+			}
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = *deprecatedVersion
+			continue
+		}
+
+		// only keep one supported version
+		if version.Classification != nil && *version.Classification == gardencorev1beta1.ClassificationSupported {
+			key := helper.GetSupportedVersionsKey(*currentSemverVersion)
+			currentlyHighestSupportedVersion, found := supportedVersionsMap[key]
+			if !found {
+				supportedVersionsMap[key] = helper.HighestSupportedVersion{
+					Version: currentSemverVersion,
+					Index:   cloudProfileIndex,
+				}
+				continue
+			}
+			if currentlyHighestSupportedVersion.Version.LessThan(currentSemverVersion) {
+				deprecatedVersion, err := deprecateKubernetesVersion(logger, config, profile, profile.Spec.Kubernetes.Versions[currentlyHighestSupportedVersion.Index], *currentSemverVersion, true, false)
+				if err != nil {
+					return false, nil, err
+				}
+				profile.Spec.Kubernetes.Versions[currentlyHighestSupportedVersion.Index] = *deprecatedVersion
+				supportedVersionsMap[key] = helper.HighestSupportedVersion{
+					Version: currentSemverVersion,
+					Index:   cloudProfileIndex,
+				}
+				continue
+			}
+
+			// deprecate current version, as there is already another higher supported version of the same minor
+			deprecatedVersion, err := deprecateKubernetesVersion(logger, config, profile, version, *currentSemverVersion, true, false)
+			if err != nil {
+				return false, nil, err
+			}
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = *deprecatedVersion
+			continue
+		}
+
+		// expire deprecated version
+		isExpired := version.ExpirationDate != nil && (time.Now().UTC().After(version.ExpirationDate.Time) || time.Now().UTC().Equal(version.ExpirationDate.Time))
+		if isExpired && version.Classification != nil && *version.Classification == gardencorev1beta1.ClassificationDeprecated {
+			logger.Infof("[KUBERNETES VERSION MANAGEMENT] expiring kubernetes version '%s'", version.Version)
+			profile.Spec.Kubernetes.Versions[cloudProfileIndex] = helper.SetExpiredClassification(version)
```

Comment:
I thought "expired" is only a logical state that is implicitly set when the expiration date is in the past. Now we have made it an actual state, right? How do you ensure that you update the classification as soon as the expiration date of any version expires?

Reply:
I made it to be a logical state, as having it as an actual state causes too many problems. Thanks!